profiles/killbill/src/test/java/org/killbill/billing/jaxrs/TestPayment.java

```diff
@@ -219,6 +219,50 @@ public void testWithTimeoutPayment() throws Exception {
+        Assert.assertEquals(initialPayments.size(), 1);
+        Assert.assertEquals(initialPayments.get(0).getTransactions().size(), 1);
+        Assert.assertEquals(initialPayments.get(0).getTransactions().get(0).getStatus(), TransactionStatus.PAYMENT_FAILURE);
+        Assert.assertEquals(initialPayments.get(0).getPaymentAttempts().size(), 2);
+        Assert.assertEquals(initialPayments.get(0).getPaymentAttempts().get(0).getStateName(), "RETRIED");
+        Assert.assertEquals(initialPayments.get(0).getPaymentAttempts().get(1).getStateName(), "SCHEDULED");
+        final InvoicePayments initialInvoicePayments = invoiceApi.getPaymentsForInvoice(failedInvoice.getInvoiceId(), requestOptions);
+        Assert.assertEquals(initialInvoicePayments.size(), 1);
+        Assert.assertEquals(initialInvoicePayments.get(0).getPaymentId(), initialPayments.get(0).getPaymentId());
+
+        // Trigger manually an invoice payment but make the control plugin abort it
+        mockPaymentControlProviderPlugin.setAborted(true);
+        final HashMultimap<String, String> queryParams = HashMultimap.create();
+        queryParams.putAll("controlPluginName", Arrays.<String>asList(MockPaymentControlProviderPlugin.PLUGIN_NAME));
+        final RequestOptions inputOptions = RequestOptions.builder()
+                                                          .withCreatedBy(createdBy)
+                                                          .withReason(reason)
+                                                          .withComment(comment)
+                                                          .withQueryParams(queryParams).build();
+        final InvoicePayment invoicePayment = new InvoicePayment();
+        invoicePayment.setPurchasedAmount(failedInvoice.getBalance());
+        invoicePayment.setAccountId(failedInvoice.getAccountId());
+        invoicePayment.setTargetInvoiceId(failedInvoice.getInvoiceId());
+        final InvoicePayment invoicePaymentNull = invoiceApi.createInstantPayment(failedInvoice.getInvoiceId(), invoicePayment, true, null, inputOptions);
+        Assert.assertNull(invoicePaymentNull);
+
+        // Verify new state
+        final Payments updatedPayments = accountApi.getPaymentsForAccount(account.getAccountId(), true, false, ImmutableMap.<String, String>of(), AuditLevel.NONE, requestOptions);
+        Assert.assertEquals(updatedPayments.size(), 1);
+        Assert.assertEquals(updatedPayments.get(0).getPaymentId(), initialPayments.get(0).getPaymentId());
+        final InvoicePayments updatedInvoicePayments = invoiceApi.getPaymentsForInvoice(failedInvoice.getInvoiceId(), requestOptions);
+        Assert.assertEquals(updatedInvoicePayments.size(), 1);
+        Assert.assertEquals(updatedInvoicePayments.get(0).getPaymentId(), updatedPayments.get(0).getPaymentId());
```

Comment:
If I understand correctly this returns the first payment that failed (with a pending retry), and the one made manually that got aborted is correctly discarded from the list -- and before it was causing the NPE ?